query-engine/core/src/interpreter/query_interpreters/write.rs

```diff
@@ -43,9 +43,16 @@ async fn create_one(
     q: CreateRecord,
     trace_id: Option<String>,
 ) -> InterpretationResult<QueryResult> {
-    let res = tx.create_record(&q.model, q.args, trace_id).await?;
+    let res = tx.create_record(&q.model, q.args, q.selected_fields, trace_id).await?;
 
-    Ok(QueryResult::Id(Some(res)))
+    Ok(QueryResult::RecordSelection(Box::new(RecordSelection {
```

Comment:
How is that changing the semantics of the result from `QueryResult::Id` to `QueryResult::RecordSelection` don't break any internal contract? Aren't there used or eventually serialized different? if so, why do we have the two different enum variants in `QueryResult`?

Reply:
Both `QueryResult::Id` and `QueryResult::RecordSelection` can be used as results. Have a look at this code [here](https://github.com/prisma/prisma-engines/blob/6b66709c6adc3189ef95f9dfb4488538af0ab108/query-engine/core/src/interpreter/expressionista.rs#L355-L363) and [here](https://github.com/prisma/prisma-engines/blob/a651a3274a6dd554823daff4f94fccb1c99b7ac9/query-engine/core/src/interpreter/interpreter_impl.rs#L54-L74), where we extract results out of both.

I wasn't here when these abstractions were created so I can only assume why we have both. `QueryResult::RecordSelection` needs a lot more information as it ends up being serialized. `QueryResult::Id` is just a simplified version of `RecordSelection` when we know it's only used as a temporary result for another operation (eg: create --(create id)--> read).

This was holding well until now, but we're stretching the abstraction by making create's result now sometimes go through the serialization (in an atomic create), and sometimes not (when we still need a read after the create).

As said below, I could make `create_record` return a `RecordSelection` or an `Id` based on the type of create we're doing, but I didn't think it was worth it as we're perfectly able to extract the result out of both `Id` and `RecordSelection`. I do agree that this is all a bit fuzzy atm. It's often the case when slapping new concepts on top of existing abstractions without really evolving them. The same questions are gonna be relevant when I tackle atomic update, which works the exact same way.

My thoughts atm is that the abstractions still covers our need but we might have to change them a bit in the future if we add more subtleties to operations. Until then, I didn't want to prematurely introduce more concepts. With that being said, I'm 100% open to suggestions